controllers/drplacementcontrol_controller.go

```diff
@@ -1602,65 +1742,224 @@ func (s ObjectStorePVDownloader) DownloadPVs(ctx context.Context, r client.Reade
+	if d.instance.Status.Phase != nextState {
+		d.log.Info(fmt.Sprintf("Phase: Current '%s'. Next '%s'",
+			d.instance.Status.Phase, nextState))
+
+		d.instance.Status.Phase = nextState
+		d.updateConditions()
+		d.needStatusUpdate = true
+	}
+}
+
+func (d *DRPCInstance) updateConditions() {
+	d.log.Info(fmt.Sprintf("Current Conditions '%v'", d.instance.Status.Conditions))
+
+	for _, condType := range []string{rmn.ConditionAvailable, rmn.ConditionReconciling} {
+		condition := d.newCondition(condType)
 
-	d.setDRState(rmn.DRState(""))
+		idx, _ := GetDRPCCondition(&d.instance.Status, condType)
+		if idx == -1 {
+			d.instance.Status.Conditions = append(d.instance.Status.Conditions, *condition)
+		} else {
+			d.instance.Status.Conditions[idx] = *condition
+		}
+	}
+
+	d.log.Info(fmt.Sprintf("Updated Conditions '%v'", d.instance.Status.Conditions))
 }
 
-func (d *DRPCInstance) setDRState(nextState rmn.DRState) {
-	if d.instance.Status.LastKnownDRState != nextState {
-		d.log.Info(fmt.Sprintf("LastKnownDRState: Current State '%s'. Next State '%s'",
-			d.instance.Status.LastKnownDRState, nextState))
+func (d *DRPCInstance) newCondition(condType string) *metav1.Condition {
+	return &metav1.Condition{
+		Type:   condType,
+		Status: d.getConditionStatus(condType),
+		// ObservedGeneration: d.getObservedGeneration(condType),
+		LastTransitionTime: metav1.Now(),
+		Reason:             d.getConditionReason(condType),
+		Message:            d.getConditionMessage(condType),
+	}
+}
 
-		d.instance.Status.LastKnownDRState = nextState
-		d.needStatusUpdate = true
+// func (d *DRPCInstance) getConditionType() string {
+// 	if d.instance.Status.Phase == "" {
+// 		return rmn.ConditionNotAvailable
+// 	}
+
+// 	return rmn.ConditionAvailable
+// }
+
+func (d *DRPCInstance) getConditionStatus(condType string) metav1.ConditionStatus {
+	if condType == rmn.ConditionAvailable {
+		return d.getConditionStatusForTypeAvailable()
+	} else if condType == rmn.ConditionReconciling {
+		return d.getConditionStatusForTypeReconciling()
+	}
+
+	return metav1.ConditionUnknown
+}
+
+func (d *DRPCInstance) getConditionStatusForTypeAvailable() metav1.ConditionStatus {
+	if d.isInFinalPhase() {
+		return metav1.ConditionTrue
+	}
+
+	if d.isInProgressingPhase() {
+		return metav1.ConditionFalse
 	}
+
+	return metav1.ConditionUnknown
 }
 
+func (d *DRPCInstance) getConditionStatusForTypeReconciling() metav1.ConditionStatus {
+	if d.isInFinalPhase() {
+		if d.reconciling() {
+			return metav1.ConditionFalse
+		}
+
+		return metav1.ConditionTrue
+	}
+
+	if d.isInProgressingPhase() {
+		return metav1.ConditionFalse
+	}
+
+	return metav1.ConditionUnknown
+}
+
+func (d *DRPCInstance) reconciling() bool {
+	idx, condition := GetDRPCCondition(&d.instance.Status, rmn.ConditionReconciling)
+	d.log.Info(fmt.Sprintf("Condition %v", condition))
+
+	if idx == -1 {
+		return false
+	}
+
+	if condition.Reason == rmn.ReasonSuccess && condition.Status == metav1.ConditionTrue {
```

Comment:
We should check if condition generation is the same as spec generation here. An older generation may update this condition to True, subsequent generation needs to start the condition fresh, which is handled using the generation numbers.